### PR TITLE
CPS CAF fetching integration

### PR DESCRIPTION
**Pull request by amoore108**
**Overall change:**
- This is a rough implementation to test fetching/rendering CPS Story assets in our ArticlePage component to try and consolidate both our fetching and rendering mechanism
- Makes use of the existing `renderer_env=` query param, to let the server know which endpoint to fetch from and in turn which page component to render
- This is intended to allow more easy comparison between rendering STY pages in our StoryPage and ArticlePage components. There will most likely be visual glitches, missing data and broken pages when using `renderer_env=caftest` or `renderer_env=caflive`!
- This will enable further spike work into finding the nuances that we need to fix    

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [ ] I have assigned this PR to the Simorgh project
- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/lib/utilities/ichefURL/index.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ const buildIChefURL = ({
     'news',
     isPng ? 'png' : null,
     resolution,
-    originCode,
+    !locator.includes(originCode) ? originCode : null,
```
Comment (amoore108) [inline src/app/lib/utilities/ichefURL/index.js:36]: Some CPS assets coming through the CAF had the `originCode` value within the `locator` field. The existing CPS data transformation logic may have been doing stuff to mitigate this, so I just added a simple check here to fix the problem.

File: src/app/pages/ArticlePage/PagePromoSections/TopStoriesSection/TopStoriesItem/index.jsx [hunk 2]
```diff
@@ -62,6 +62,7 @@ const TopStoriesItem = forwardRef(
               <Promo.Link>
                 {isLive ? (
                   <Promo.LiveLabel
+                    service={service}
```
Comment (amoore108) [inline src/app/pages/ArticlePage/PagePromoSections/TopStoriesSection/TopStoriesItem/index.jsx:65]: Fixes a prop-type error

File: src/app/routes/article/getInitialData/index.ts [hunk 3]
```diff
@@ -23,6 +25,11 @@ const getEnvironment = (pathname: string) => {
   if (pathname.includes('renderer_env=live')) {
     return 'live';
   }
+  if (pathname.includes('renderer_env=caf')) {
+    return process.env.SIMORGH_APP_ENV !== 'local'
+      ? process.env.SIMORGH_APP_ENV
+      : 'live';
+  }
```
Comment (amoore108) [inline src/app/routes/article/getInitialData/index.ts:32]: If `renderer_env=caf` is supplied, it will set the environment to whatever environment Simorgh is running in **if its not running locally**, in which case it will default to `live`.

This is to set the `ctx-service-env` header for fetching from the BFF. Seemed to make more sense to test against Live assets since the metadata and page data will be much more accurate for comparison. 

You will need to manually set the BFF url to point to your locally running instance of FABL if you want to test FABL locally.

File: src/app/routes/article/getInitialData/index.ts [hunk 4]
```diff
@@ -14,13 +13,6 @@ import getOnwardsPageData from '../utils/getOnwardsData';
 
 const logger = nodeLogger(__filename);
 
-const removeAmp = (path: string) => path.split('.')[0];
-const getOptimoId = (path: string) => path.match(/(c[a-zA-Z0-9]{10}o)/)?.[1];
-const getCpsId = (path: string) => path;
-
-const getId = (pageType: string) =>
-  pipe(getUrlPath, removeAmp, pageType === 'article' ? getOptimoId : getCpsId);
```
Comment (amoore108) [inline src/app/routes/article/getInitialData/index.ts:22]: This logic is now in the BFF module https://github.com/bbc/fabl-modules/pull/3740/commits/d08adb395f50b83f3cc828411c6803997d369e87

File: src/app/routes/utils/fetchPageData/utils/getRouteProps/index.js [hunk 5]
```diff
@@ -34,12 +34,17 @@ const getRouteProps = url => {
...
+      : undefined;
+
   return {
     isAmp: amp || isAmpPath(url),
     isApp: app || isAppPath(url),
     service: service || fallbackServiceParam(url),
     variant,
-    id,
+    id: id || cpsId,
     assetUri,
```
Comment (amoore108) [inline src/app/routes/utils/fetchPageData/utils/getRouteProps/index.js:48]: I've combined `id` which comes from Optimo/TIPO routes and `cpsId` which is CPS routes.

The `id` field is only used in the `ArticleMediaPlayer` component from what I can see: https://github.com/bbc/simorgh/blob/cf3ce02abe7e035805133551e91f9822fa25a223/src/app/legacy/containers/ArticleMediaPlayer/index.jsx#L11 We need to set it to either the Optimo or CPS ID for the media player to work correctly when rendering both page types in the same page component.

`assetUri` is set, but doesn't appear to be used at all, or at least isn't being passed into our Contexts: https://github.com/bbc/simorgh/blob/cf3ce02abe7e035805133551e91f9822fa25a223/src/app/legacy/containers/PageHandlers/withContexts.jsx so it could probably be removed or kept to differentiate between Optimo/TIPO and CPS IDs, although the naming isn't very clear.

Comment (amoore108) [reply src/app/routes/utils/fetchPageData/utils/getRouteProps/index.js:48]: Appears to be straightforward to combine them, just requires updating quite a few references and tests that set the `id` field manually.

File: src/app/legacy/containers/ArticleMediaPlayer/index.jsx [hunk 6]
```diff
@@ -7,15 +7,30 @@ import {
 } from '#models/propTypes';
 import MediaPlayerContainer from '../MediaPlayer';
 
+const formatAssetId = assetUri => {
+  if (assetUri && assetUri.includes('newyddion')) {
+    return assetUri.replace('newyddion', 'cymrufyw');
```
Comment (andrewscfc) [inline src/app/legacy/containers/ArticleMediaPlayer/index.jsx:12]: Why is that needed for this change?

File: src/app/legacy/containers/ArticleMediaPlayer/index.jsx [hunk 7]
```diff
@@ -7,15 +7,30 @@ import {
...
+  const { id, pageType } = useContext(RequestContext);
 
   return (
     <GridItemMediumNoMargin>
       <MediaPlayerContainer
         blocks={blocks}
-        assetId={id}
-        assetType="articles"
+        assetId={formatAssetId(id)}
+        assetType={pageTypeMap[pageType]}
```
Comment (andrewscfc) [inline src/app/legacy/containers/ArticleMediaPlayer/index.jsx:33]: what is assetType used for?

File: src/app/models/propTypes/metadata/index.js [hunk 8]
```diff
@@ -35,7 +35,8 @@ export const optimoMetadataPropTypes = {
   lastPublished: number.isRequired,
   lastUpdated: number,
   locators: shape({
-    optimoUrn: string.isRequired,
+    optimoUrn: string,
+    cpsUrn: string,
```
Comment (andrewscfc) [inline src/app/models/propTypes/metadata/index.js:39]: Does this facilitate a CPS Urn on an article page?

File: src/app/pages/ArticlePage/ArticlePage.jsx [hunk 9]
```diff
@@ -235,7 +236,8 @@ const ArticlePage = ({ pageData }) => {
               tagBackgroundColour={WHITE}
             />
           )}
-          <RelatedContentSection content={blocks} />
+          {/* TODO: Related Content section needs special formatting of CPS assets when using CAF endpoint */}
+          {!isCaf && <RelatedContentSection content={blocks} />}
```
Comment (andrewscfc) [inline src/app/pages/ArticlePage/ArticlePage.jsx:240]: I guess we are considering not supporting related content with CAF?

File: src/app/routes/cpsAsset/getInitialData/index.js [hunk 10]
```diff
@@ -137,7 +139,7 @@ export default async ({
     const response = {
       status,
       pageData: {
-        ...(await transformJson(article, pathname, toggles)),
+        ...(isCaf ? article : await transformJson(article, pathname, toggles)),
```
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/index.js:142]: I guess this is the cruical point where we could delete everything below `transformJson` in the call stack if made this permanent 🥳 ?

**Review by andrewscfc — APPROVED** [review]

File: src/app/legacy/containers/ArticleMediaPlayer/index.jsx (hunk 6 above)
Comment (amoore108) [reply src/app/legacy/containers/ArticleMediaPlayer/index.jsx:12]: This was added from seeing the logic in the CpsAssetMediaPlayer component:
https://github.com/bbc/simorgh/blob/d03d06e0a0e3025b46c1546d297a3f4da93ff0e9/src/app/legacy/containers/CpsAssetMediaPlayer/index.jsx#L80

https://github.com/bbc/simorgh/blob/c9253b928bc4db268c1d3eff237657a53c777633/src/app/legacy/containers/CpsAssetMediaPlayer/utils/getAssetUri/index.js#L0-L1

We have similar logic in the BFF when fetching for Welsh assets to convert the language name. This is just required to get the media player to work correctly in the Welsh language when routing CPS assets through this component.

File: src/app/legacy/containers/ArticleMediaPlayer/index.jsx (hunk 7 above)
Comment (amoore108) [reply src/app/legacy/containers/ArticleMediaPlayer/index.jsx:33]: `assetType` is passed down to the function that constructs the `ws-embeds` URL:
https://github.com/bbc/simorgh/blob/655513fb4ddef89611aad86c071073e3c41ded65/src/app/lib/utilities/getUrlHelpers/getEmbedUrl/index.js#L50

So it will either have `/articles/` or `/cps/` in the route. Just another compatibility thing to use the ArticleMediaPlayer for both Optimo and CPS assets.

File: src/app/models/propTypes/metadata/index.js (hunk 8 above)
Comment (amoore108) [reply src/app/models/propTypes/metadata/index.js:39]: Trying to remember why I added this tbh. I think it was just to get rid of a prop-type warning in the console as the `locators` object has those 2 different fields depending on whether its an Optimo or CPS asset.

File: src/app/pages/ArticlePage/ArticlePage.jsx (hunk 9 above)
Comment (amoore108) [reply src/app/pages/ArticlePage/ArticlePage.jsx:240]: Its something to think about. We can easily enough add a data transform in the BFF to handle the Optimo specific structure of related content. Just doesn't seem to be something CAF automatically does. Yet at least.

Or, arguably, we could just not show related content for CPS assets as they will be deemed 'legacy/archive', so perhaps related content and other onward-journey elements are less needed there.

File: src/app/routes/cpsAsset/getInitialData/index.js (hunk 10 above)
Comment (amoore108) [reply src/app/routes/cpsAsset/getInitialData/index.js:142]: Yep thats right! The entire `transformJson` and all those Optimo processing bits can be removed.

In fact, this entire file could be removed for STY/CSP page types and we would just use the Article `getIinitalData` function on its own and remove `cpsAsset` as a possible `pageType`: https://github.com/bbc/simorgh/blob/a2797f2b803016dc12d9d6bd4f60d7fbd7c8ab59/src/app/routes/article/getInitialData/index.ts#L21

Comment (andrewscfc) [reply src/app/routes/cpsAsset/getInitialData/index.js:142]: Really want this, we would finally have no explicit support for CPS ❤️

**Review by shayneahchoon — APPROVED** [review]
